apps/language_server/lib/language_server/source_file.ex

```diff
@@ -63,31 +100,44 @@ defmodule ElixirLS.LanguageServer.SourceFile do
 
   def full_range(source_file) do
     lines = lines(source_file)
+    last_line = List.last(lines)
+
+    utf16_size =
+      :unicode.characters_to_binary(last_line, :utf8, :utf16)
+      |> byte_size()
+      |> div(2)
 
     %{
       "start" => %{"line" => 0, "character" => 0},
-      "end" => %{"line" => Enum.count(lines) - 1, "character" => String.length(List.last(lines))}
+      "end" => %{"line" => Enum.count(lines) - 1, "character" => utf16_size}
     }
   end
 
+  defp prepend_line(line, nil, acc), do: [line | acc]
+  defp prepend_line(line, ending, acc), do: [[line, ending] | acc]
+
   def apply_edit(text, range(start_line, start_character, end_line, end_character), new_text) do
     lines_with_idx =
       text
-      |> lines()
+      |> lines_with_endings()
       |> Enum.with_index()
 
     acc =
-      Enum.reduce(lines_with_idx, [], fn {line, idx}, acc ->
+      Enum.reduce(lines_with_idx, [], fn {{line, ending}, idx}, acc ->
         cond do
           idx < start_line ->
-            [[line, ?\n] | acc]
+            prepend_line(line, ending, acc)
 
           idx == start_line ->
             # LSP contentChanges positions are based on UTF-16 string representation
             # https://microsoft.github.io/language-server-protocol/specification#textDocuments
             beginning_utf8 =
               :unicode.characters_to_binary(line, :utf8, :utf16)
-              |> binary_part(0, start_character * 2)
+              |> (&binary_part(
+                    &1,
+                    0,
+                    min(start_character * 2, byte_size(&1))
```

Comment:
Can you explain why `min` is needed here? And why `min/max` is needed for the other similar call later in the file. I'm thinking that it would be better to extract a helper function since this function is already quite large and indexing into a utf8 string as if it was utf16 is a nice well-contained problem. I took a stab at doing it myself, but I wasn't clear on the `min/max` constraints so I couldn't integrate those well.

Reply:
Without that it crashes with ArgumentError when range is invalid (starts at negative, starts over length or takes more chars than are in binary). It shouldn't happen normally but it does when client looses synch with the server, e.g. in https://github.com/elixir-lsp/elixir-ls/pull/277